### PR TITLE
Add incremental training and related unit tests

### PR DESCRIPTION
**Pull request by yunboouyang**
Changes are:
(1) Add `PriorDistribution` to implement incremental learning prior distribution loss function
(2) Enable photon-ml to read variance vectors
(3) Add 2 new parameters: i. Boolean `incrementalLearning` as the switch for incremental learning. ii. `incrementalWeight` to specify the importance of prior distribution.

Current implementation only covers diagonal prior variance. For Full Hessian implementation, see https://github.com/linkedin/photon-ml/pull/450

### CODE REVIEW

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+   * @return The Hessian diagonal of the Gaussian regularization term
+   */
+  protected def l2RegHessianDiagonal: Vector[Double] = l2RegWeight * inversePriorVariances
+
+  /**
+   * Compute the Hessian matrix of the Gaussian regularization term for the given model coefficients.
+   *
+   * @return The Hessian matrix of the Gaussian regularization term
+   */
+  protected def l2RegHessianMatrix: DenseMatrix[Double] = l2RegWeight * diag(inversePriorVariances)
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: Thank you for catching some mistakes in my implementation of the mathematics.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    assertEquals(mockObjectiveFunction.value(Unit, coefficients, mockNormalization), expectedValue)
+    assertEquals(mockObjectiveFunction.gradient(Unit, coefficients, mockNormalization), expectedGradient)
+    assertEquals(
+      mockObjectiveFunction.hessianVector(Unit, coefficients, multiplyVector, mockNormalization),
+      expectedVector)
+    assertEquals(mockObjectiveFunction.hessianDiagonal(Unit, coefficients), expectedDiagonal)
+    assertEquals(mockObjectiveFunction.hessianMatrix(Unit, coefficients), expectedMatrix)
+  }
+}
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala]: Missing newline character at the end of file

File: photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    val l2Weight = 10D
+
+    val mockObjectiveFunction = new MockObjectiveFunction with PriorDistributionTwiceDiff {
+      override val priorCoefficients = ModelCoefficients(priorMean, Option(priorVar))
+      l1RegWeight = l1Weight
+      l2RegWeight = l2Weight
+    }
+
+    // Assume that coefficients = 1-vector, prior mean = 2-vector, multiply = 3-vector, prior variance = 4-vector for all expected values below
+    val expectedValue = MockObjectiveFunction.VALUE + 0.25 * l2Weight * DIMENSION / 2 + 0.5 * l1Weight * DIMENSION
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala]: You might want to add some comments explaining the mathematics, or do more computations with the raw variables. There's some skipped steps/simplified values in these calculations which confused me for a bit, even though I'm looking at the design document.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala (hunk 1 above)
Comment (yunboouyang) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: fixed

File: photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala [hunk 4]
```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2018 LinkedIn Corp. All rights reserved.
```
[the remark this answers is not in the document]
Comment (yunboouyang) [reply photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala]: fixed

File: photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala (hunk 2 above)
Comment (yunboouyang) [reply photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala]: fixed

File: photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala (hunk 3 above)
Comment (yunboouyang) [reply photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala]: Fixed. Add more comments in `PriorDistribution.scala` and `PriorDistributionTest.scala`

File: photon-api/src/integTest/scala/com/linkedin/photon/ml/algorithm/CoordinateFactoryIntegTest.scala [hunk 5]
```diff
@@ -139,7 +146,7 @@ object CoordinateFactoryIntegTest {
   private val INTERCEPT_INDEX = None
 
   private val OPTIMIZER_CONFIG = OptimizerConfig(OPTIMIZER_TYPE, MAX_ITER, TOLERANCE)
-  private val MOCK_NORMALIZATION = mock(classOf[NormalizationContext])
+  private val MOCK_NORMALIZATION = mock(classOf[NormalizationContext], withSettings().serializable())
```
Comment (ashelkovnykov) [inline photon-api/src/integTest/scala/com/linkedin/photon/ml/algorithm/CoordinateFactoryIntegTest.scala]: What was the reason for this change?

File: photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    val mockObjectiveFunction = new MockObjectiveFunction with PriorDistributionTwiceDiff {
+      override val priorCoefficients = ModelCoefficients(priorMean, Option(priorVar))
+      l1RegWeight = l1Weight
+      l2RegWeight = l2Weight
+    }
+
+    /**
+     * Assume that coefficients = 1-vector, prior mean = 2-vector, multiply = 3-vector, prior variance = 4-vector for all expected values below
+     * l2RegValue = pow(1 - 2, 2) / 4 * l2Weight * DIMENSION / 2 = 0.25 * l2Weight * DIMENSION / 2;
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala]: `DIMENSION` could still use a better explanation: we're taking the L2 norm of the difference between prior and new coefficients, which for this toy case is the 1-vector. L2 norm of the 1-vector is `d` where `d` is the dimension of the vector.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+     * l1RegValue = abs(1 - 2) / 2 * l2Weight * DIMENSION = 0.5 * l2Weight * DIMENSION;
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala]: `DIMENSION` could still use a better explanation: we're taking the L1 norm of the difference between prior and new coefficients, which for this toy case is the 1-vector. L1 norm of the 1-vector is `d` where `d` is the dimension of the vector.

File: photon-api/src/integTest/scala/com/linkedin/photon/ml/algorithm/CoordinateFactoryIntegTest.scala (hunk 5 above)
Comment (yunboouyang) [reply photon-api/src/integTest/scala/com/linkedin/photon/ml/algorithm/CoordinateFactoryIntegTest.scala]: Integ test throws an error that  `MOCK_NORMALIZATION` is not serializable. Original class `NormalizatoinContext` extends `Serializable`. Need to set `MOCK_NORMALIZATION` to be serializable as well.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala (hunk 7 above)
Comment (yunboouyang) [reply photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala]: fixed

File: photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala (hunk 6 above)
Comment (yunboouyang) [reply photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala]: fixed

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+  /**
+   * Compute the gradient of the Gaussian regularization term for the given model coefficients. Gradient is
+   * l2RegWeight * (coefficients - priorMeans) :/ priorVariance.
+   *
+   * @param coefficients The model coefficients
+   * @return The gradient of the Gaussian regularization term
+   */
+  protected def l2RegGradient(coefficients: Vector[Double]): Vector[Double] = {
+
+    val normalizedCoefficients = (coefficients - priorMeans) *:* inversePriorVariances
```
Comment (rrohan) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: This is one point of difference that we can discuss.

File: photon-api/src/main/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblem.scala [hunk 9]
```diff
@@ -87,10 +89,7 @@ protected[ml] class DistributedOptimizationProblem[Objective <: DistributedObjec
 
     val result = (objectiveFunction, varianceComputation) match {
```
Comment (rrohan) [inline photon-api/src/main/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblem.scala]: If this is where you are updating the variance, this is another point of difference

File: photon-api/src/main/scala/com/linkedin/photon/ml/optimization/SingleNodeOptimizationProblem.scala [hunk 10]
```diff
@@ -59,9 +59,7 @@ protected[ml] class SingleNodeOptimizationProblem[Objective <: SingleNodeObjecti
   override def computeVariances(input: Iterable[LabeledPoint], coefficients: Vector[Double]): Option[Vector[Double]] =
     (objectiveFunction, varianceComputationType) match {
       case (twiceDiffFunc: TwiceDiffFunction, VarianceComputationType.SIMPLE) =>
-        Some(twiceDiffFunc
-          .hessianDiagonal(input, coefficients)
-          .map(v => 1.0 / math.max(v, MathConst.EPSILON)))
+        Some(VectorUtils.invertVector(twiceDiffFunc.hessianDiagonal(input, coefficients)))
```
Comment (rrohan) [inline photon-api/src/main/scala/com/linkedin/photon/ml/optimization/SingleNodeOptimizationProblem.scala]: And this too, if random effects are updated here

File: photon-api/src/main/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblem.scala (hunk 9 above)
Comment (yunboouyang) [reply photon-api/src/main/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblem.scala]: Yes. photon-ml will output Hessian diagonal. To output variance, I only need to invert it.  This is for fixed effect model variance update.

File: photon-api/src/main/scala/com/linkedin/photon/ml/optimization/SingleNodeOptimizationProblem.scala (hunk 10 above)
Comment (yunboouyang) [reply photon-api/src/main/scala/com/linkedin/photon/ml/optimization/SingleNodeOptimizationProblem.scala]: Yes. This is for random effect model variance update.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala (hunk 8 above)
Comment (yunboouyang) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: We use the prior model information by adding a new regularization term: (coefficients - priorMeans) ^ 2 / priorVariances * l2RegWeight / 2

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+   * Compute the Gaussian regularization term for the given model coefficients. L2 regularization term is
+   * l2RegWeight * sum(pow(coefficients - priorMeans, 2) :/ priorVariance) / 2.
+   *
+   * @param coefficients The model coefficients
+   * @return The Gaussian regularization term value
+   */
+  protected def l2RegValue(coefficients: Vector[Double]): Double = {
+
+    val normalizedCoefficients = (coefficients - priorMeans) *:* sqrt(inversePriorVariances)
+
```
Comment (junshi15) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: why sqrt, then square later? why not multiple it  at line 98

Comment (yunboouyang) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: These are equivalent. We can merge the 2 lines together

File: photon-api/src/main/scala/com/linkedin/photon/ml/data/RandomEffectDataset.scala [hunk 12]
```diff
@@ -392,12 +396,26 @@ object RandomEffectDataset {
...
+      .mapValues { case (projector: LinearSubspaceProjector, priorModelOpt: Option[GeneralizedLinearModel]) =>
+        val activeCoef = priorModelOpt.map {
+            model =>
+              val means = model.coefficients.means
+              VectorUtils.getActiveIndices(means)
+        }.getOrElse(Set[Int]())
+        val projectedKeySet = projector.originalToProjectedSpaceMap.keySet
+
+        new LinearSubspaceProjector(activeCoef.union(projectedKeySet).toSet, originalSpaceDimension)
+      }
```
Comment (ashelkovnykov) [inline photon-api/src/main/scala/com/linkedin/photon/ml/data/RandomEffectDataset.scala]: There's a much cleaner way to write this function:
```
    val sc = SparkSession.builder().getOrCreate().sparkContext
    val originalSpaceDimension = keyedGameDataset
      .take(1)
      .head
      ._2
      ._2
      .features
      .length
    val priorModels = priorRandomEffectModelOpt.map(_.modelsRDD).getOrElse(sc.emptyRDD[(REId, GeneralizedLinearModel)])
    
    keyedGameDataset
      .leftOuterJoin(priorModels)
      .mapValues { case ((_, labeledPoint), priorModelOpt) =>
        val dataActiveIndices = VectorUtils.getActiveIndices(labeledPoint.features)
          
        priorModelOpt match {
          case Some(priorModel) => dataActiveIndices.union(VectorUtils.getActiveIndices(priorModel.coefficients.means))
          case None => dataActiveIndices
        }
      }
      .foldByKey(mutable.Set[Int](), randomEffectPartitioner)(_.union(_))
      .mapValues(activeIndices => new LinearSubspaceProjector(activeIndices.toSet, originalSpaceDimension))
```

File: photon-api/src/main/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblem.scala [hunk 13]
```diff
@@ -28,7 +28,7 @@ import com.linkedin.photon.ml.optimization.VarianceComputationType.VarianceCompu
 import com.linkedin.photon.ml.optimization.game.GLMOptimizationConfiguration
 import com.linkedin.photon.ml.sampling.DownSampler
 import com.linkedin.photon.ml.supervised.model.GeneralizedLinearModel
-import com.linkedin.photon.ml.util.BroadcastWrapper
+import com.linkedin.photon.ml.util.{BroadcastWrapper, VectorUtils}
```
Comment (ashelkovnykov) [inline photon-api/src/main/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblem.scala]: `MathConst` import is unused

File: photon-api/src/main/scala/com/linkedin/photon/ml/optimization/SingleNodeOptimizationProblem.scala [hunk 14]
```diff
@@ -24,7 +24,7 @@ import com.linkedin.photon.ml.normalization.NormalizationContext
 import com.linkedin.photon.ml.optimization.VarianceComputationType.VarianceComputationType
 import com.linkedin.photon.ml.optimization.game.GLMOptimizationConfiguration
 import com.linkedin.photon.ml.supervised.model.GeneralizedLinearModel
-import com.linkedin.photon.ml.util.BroadcastWrapper
+import com.linkedin.photon.ml.util.{BroadcastWrapper, VectorUtils}
```
Comment (ashelkovnykov) [inline photon-api/src/main/scala/com/linkedin/photon/ml/optimization/SingleNodeOptimizationProblem.scala]: `MathConst` import is unused

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroUtils.scala [hunk 15]
```diff
@@ -422,7 +445,7 @@ object AvroUtils {
    * @return The (effectId, latentFactor) pair converted from the input Avro record
    */
   protected[avro] def convertLatentFactorAvroToLatentFactor(
-      latentFactorAvro: LatentFactorAvro): (String, Vector[Double]) = {
+    latentFactorAvro: LatentFactorAvro): (String, Vector[Double]) = {
```
Comment (ashelkovnykov) [inline photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroUtils.scala]: Undo indent change

File: photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroUtils.scala [hunk 16]
```diff
@@ -399,6 +392,36 @@ object AvroUtils {
...
+  /**
+   * Convert the NameTermValueAvro List of the type [[JList[NameTermValue]]] to Breeze vector of type [[Vector[Double]]].
+   *
+   * @param nameTermValueAvroList List of the type [[JList[NameTermValue]]]
+   * @param featureMap The map from feature name of type [[NameAndTerm]] to feature index of type [[Int]]
+   * @return Breeze vector of type [[Vector[Double]]]
+   */
+  protected[avro] def convertNameTermValueAvroList(
+    nameTermValueAvroList: JList[NameTermValueAvro],
+    featureMap: IndexMap): Vector[Double] = {
```
Comment (ashelkovnykov) [inline photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroUtils.scala]: Parameters should be indented once more

File: photon-lib/src/main/scala/com/linkedin/photon/ml/model/Coefficients.scala [hunk 17]
```diff
@@ -100,22 +100,22 @@ case class Coefficients(means: Vector[Double], variancesOption: Option[Vector[Do
...
-        }
-
-        sameType && sameMeans && sameVariance
-
-      case _ => false
-    }
+  override def equals(that: Any): Boolean = that match {
+    case other: Coefficients =>
+      val (m1, v1, m2, v2) = (this.means, this.variancesOption, other.means, other.variancesOption)
+      val sameType = (m1.getClass == m2.getClass) && (v1.map(_.getClass) == v2.map(_.getClass))
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/model/Coefficients.scala]: I recall we had a conversation about a bug being caused by this line, where the classes were expected to match but didn't need to. Is that not that case?

File: photon-lib/src/main/scala/com/linkedin/photon/ml/model/GameModel.scala [hunk 18]
```diff
@@ -40,8 +40,17 @@ class GameModel (private val gameModels: Map[CoordinateId, DatumScoringModel]) e
   /**
    * Get a sub-model by name.
    *
-   * @param name The model name
-   * @return An [[Option]] containing the sub-model associated with `name` in the GAME model, or `None` if none exists.
+   * @throws NoSuchElementException if no sub-model with key [[name]] exists
+   * @param name The sub-model name
+   * @return The sub-model associated with [[name]] in the GAME model
+   */
+  def apply(name: CoordinateId): DatumScoringModel = gameModels(name)
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/model/GameModel.scala]: According to IDEA, this function is unused.

File: photon-api/src/main/scala/com/linkedin/photon/ml/function/glm/SingleNodeGLMLossFunction.scala [hunk 19]
```diff
@@ -144,26 +146,43 @@ object SingleNodeGLMLossFunction {
...
+   * @param isIncrementalTrainingEnabled Is this an objective function for incremental training?
    * @return A new SingleNodeGLMLossFunction
    */
   def apply(
       configuration: GLMOptimizationConfiguration,
       singleLossFunction: PointwiseLossFunction,
-      interceptIndexOpt: Option[Int] = None): SingleNodeGLMLossFunction = {
+      priorModelOpt: Option[GeneralizedLinearModel] = None,
+      interceptIndexOpt: Option[Int] = None,
+      isIncrementalTrainingEnabled: Boolean = false): SingleNodeGLMLossFunction = {
```
Comment (ashelkovnykov) [inline photon-api/src/main/scala/com/linkedin/photon/ml/function/glm/SingleNodeGLMLossFunction.scala]: Why do we need both the `priorModelOpt` and the `isIncrementalTrainingEnabled` flag? In the match statement below, the only two valid cases are `(Some, true)` and `None, false`, meaning that if the `priorModelOpt` is present at all it implies that we're doing incremental training.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+ * regularization weight refers to the relative weight of the Gaussian prior.
+ */
+trait PriorDistribution extends ObjectiveFunction {
+
+  val priorCoefficients: ModelCoefficients = ModelCoefficients(DenseVector.zeros(1))
+
+  lazy protected val priorMeans: Vector[Double] = priorCoefficients.means
+  lazy protected val priorVariances: Vector[Double] = priorCoefficients.variancesOption.get
+  lazy protected val inversePriorVariances: DenseVector[Double] = priorVariances.map(v => if (v > MathConst.EPSILON) 1.0 / v else l2RegWeight).toDenseVector
+  protected var l2RegWeight: Double = 0D
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: What is this variable - it looks like a default regularization value for the case that some entity doesn't have a pre-existing model. Is that correct? Needs documentation.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+  protected var incrementalWeight: Double = 1D
+
+  require(l2RegWeight >= 0D, s"Invalid regularization weight '$l2RegWeight")
+
+  /**
+   * Getter for the Gaussian weight of the prior.
+   *
+   * @return The L2 regularization weight
+   */
+  def l2RegularizationWeight: Double = incrementalWeight
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: Very confusing - a function called `l2RegularizationWeight` which returns `incrementalWeight` in an object which has another member called `l2RegWeight`.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+ * Laplace distributions. The L1 regularization weight refers to the relative weight of the Laplace prior. The L2
+ * regularization weight refers to the relative weight of the Gaussian prior.
+ */
+trait PriorDistribution extends ObjectiveFunction {
+
+  val priorCoefficients: ModelCoefficients = ModelCoefficients(DenseVector.zeros(1))
+
+  lazy protected val priorMeans: Vector[Double] = priorCoefficients.means
+  lazy protected val priorVariances: Vector[Double] = priorCoefficients.variancesOption.get
+  lazy protected val inversePriorVariances: DenseVector[Double] = priorVariances.map(v => if (v > MathConst.EPSILON) 1.0 / v else l2RegWeight).toDenseVector
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: Can we find a way to consolidate this with `VectorUtils.invertVector`?

File: photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala [hunk 23]
```diff
@@ -235,25 +237,32 @@ object RandomEffectCoordinate {
...
 
               (updatedModel, Some(stateTrackers))
 
             case (localModel, _) =>
-              (localModel, None)
+              if (localModel.isDefined) {
+                (localModel.get, None)
+              } else {
+                (null, None)
+              }
```
[the remark this answers is not in the document]
Comment (yunboouyang) [reply photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala]: fixed

File: photon-api/src/main/scala/com/linkedin/photon/ml/data/RandomEffectDataset.scala (hunk 12 above)
Comment (yunboouyang) [reply photon-api/src/main/scala/com/linkedin/photon/ml/data/RandomEffectDataset.scala]: Agreed. It is much cleaner.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/model/GameModel.scala (hunk 18 above)
Comment (yunboouyang) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/model/GameModel.scala]: fixed

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala (hunk 11 above)
Comment (yunboouyang) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: fixed

File: photon-lib/src/main/scala/com/linkedin/photon/ml/model/Coefficients.scala (hunk 17 above)
Comment (yunboouyang) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/model/Coefficients.scala]: No. The bug is in https://github.com/linkedin/photon-ml/pull/445/files/d3c24eb0a934377784bc2dea1bf82e9cb308007b#diff-b0fe65697761c5d596b4067d05825968L36, where the type of the means should match the type of variance.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala (hunk 21 above)
Comment (yunboouyang) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: fixed

File: photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala (hunk 20 above)
Comment (yunboouyang) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala]: added documentation

File: photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala [hunk 24]
```diff
@@ -262,7 +264,7 @@ object RandomEffectCoordinate {
             case _ =>
               throw new IllegalStateException("Either a initial random effect model or data should be present!")
           }
-        modelsAndTrackers.persist(StorageLevel.MEMORY_ONLY_SER)
+        modelsAndTrackers.persist(StorageLevel.MEMORY_AND_DISK_SER)
```
Comment (cmjiang) [inline photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala]: Why do we need to change the StorageLevel?

File: photon-api/src/main/scala/com/linkedin/photon/ml/data/RandomEffectDataset.scala [hunk 25]
```diff
@@ -392,9 +399,16 @@ object RandomEffectDataset {
       .features
       .length
 
+    val priorModels = priorRandomEffectModelOpt.map(_.modelsRDD).getOrElse(sc.emptyRDD[(REId, GeneralizedLinearModel)])
```
Comment (cmjiang) [inline photon-api/src/main/scala/com/linkedin/photon/ml/data/RandomEffectDataset.scala]: I think these changes have been merged by PR-462. Can you merge the conflict and update the PR for an updated diff?

File: photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala [hunk 26]
```diff
@@ -244,39 +254,104 @@ class GameEstimator(val sc: SparkContext, implicit val logger: Logger) extends P
       !ignoreThreshold || initialModelOpt.isDefined,
       "'Ignore threshold for new models' flag set but no initial model provided for warm-start")
 
-    // Partial retraining and warm-start training require an initial GAME model to be provided as input
-    val coordinatesToTrain = (initialModelOpt, retrainModelCoordsOpt) match {
-      case (Some(initModel), Some(retrainModelCoords)) =>
+    // Warm-start, partial re-training, and incremental training are mutually exclusive.
```
Comment (cmjiang) [inline photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala]: Previously warm-start and partial-retraining are coupled. Now the dependency is decoupled, right?

File: photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala [hunk 27]
```diff
@@ -507,12 +583,32 @@ class GameEstimator(val sc: SparkContext, implicit val logger: Logger) extends P
             None
           }
 
-          val randomEffectDataset = RandomEffectDataset(
-            gameDataset,
-            reConfig,
-            rePartitioner,
-            existingModelKeysRddOpt,
-            StorageLevel.DISK_ONLY)
+          val randomEffectDataset = if(isIncrementalTraining) {
```
Comment (cmjiang) [inline photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala]: The following part also needs to be merged with conflicts.

File: photon-api/src/main/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblem.scala [hunk 28]
```diff
@@ -85,10 +87,7 @@ protected[ml] class DistributedOptimizationProblem[Objective <: DistributedObjec
 
     val result = (objectiveFunction, varianceComputation) match {
       case (twiceDiffFunc: TwiceDiffFunction, VarianceComputationType.SIMPLE) =>
-        Some(
-          twiceDiffFunc
-            .hessianDiagonal(input, coefficients)
-            .map(v => 1.0 / math.max(v, MathConst.EPSILON)))
+        Some(VectorUtils.invertVectorWithZeroHandler(twiceDiffFunc.hessianDiagonal(input, coefficients), 1.0 / MathConst.EPSILON))
```
Comment (cmjiang) [inline photon-api/src/main/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblem.scala]: Reformat this line.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/util/VectorUtils.scala [hunk 29]
```diff
@@ -284,4 +286,15 @@ object VectorUtils {
...
+  /**
+   * Element-wise inversion of a [[Vector]], if an element is smaller than MathConst.EPSILON, the inversion will be
+   * replaced by zeroReplacedVal.
+   *
+   * @param vector The [[Vector]] to invert
+   * @param zeroReplacedVal The replaced value when inverting an element close to 0
+   * @return The inverted [[Vector]]
+   */
+  def invertVectorWithZeroHandler(vector: Vector[Double], zeroReplacedVal: Double): Vector[Double] =
+    vector.map(v => if (v > MathConst.EPSILON) 1.0 / v else zeroReplacedVal)
```
Comment (cmjiang) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/util/VectorUtils.scala]: Shouldn't we checkout the absolute value of `v`?

File: photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala [hunk 30]
```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2019 LinkedIn Corp. All rights reserved.
```
Comment (cmjiang) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala]: Copyright 2020.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/util/VectorUtils.scala (hunk 29 above)
Comment (cmjiang) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/util/VectorUtils.scala]: Although variance is always non-negative, but I suggest to do it general for this util function.

File: photon-api/src/main/scala/com/linkedin/photon/ml/data/RandomEffectDataset.scala (hunk 25 above)
Comment (yunboouyang) [reply photon-api/src/main/scala/com/linkedin/photon/ml/data/RandomEffectDataset.scala]: updated

File: photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala (hunk 27 above)
Comment (yunboouyang) [reply photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala]: updated

File: photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala (hunk 24 above)
Comment (yunboouyang) [reply photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala]: changed back to the original storage level.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/util/VectorUtils.scala (hunk 29 above)
Comment (yunboouyang) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/util/VectorUtils.scala]: fixed

File: photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala (hunk 30 above)
Comment (yunboouyang) [reply photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala]: fixed

File: photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala (hunk 26 above)
Comment (yunboouyang) [reply photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala]: No. They are mutually exclusive. Updated the comments to make it clearer.

**Pull request merged**